pkg/runtime/wfengine/workflow.go

```diff
@@ -322,13 +322,31 @@ func (wf *workflowActor) runWorkflow(ctx context.Context, actorID string, remind
 
 	// The logic/for loop below purges/removes any leftover state from a completed or failed activity
 	// TODO: for optimization make multiple go routines and run them in parallel
-	for _, e := range state.Inbox {
+	for num, e := range state.Inbox {
 		var taskID int32
 		if ts := e.GetTaskCompleted(); ts != nil {
 			taskID = ts.TaskScheduledId
 		} else if tf := e.GetTaskFailed(); tf != nil {
 			taskID = tf.TaskScheduledId
 		} else {
+			if strings.Contains(e.String(), "timerFired") {
+				numZero := 6 - len(fmt.Sprint(num))
+				req := actors.TransactionalRequest{
+					ActorType: wf.config.workflowActorType,
+					ActorID:   actorID,
+					Operations: []actors.TransactionalOperation{{
+						Operation: actors.Upsert,
+						Request: actors.TransactionalUpsert{
+							Key: inboxKeyPrefix + "-" + strings.Repeat("0", numZero) + fmt.Sprint(num),
+						},
+					}},
+				}
+				err = wf.actors.TransactionalStateOperation(ctx, &req)
```

Comment:
Inbox updates are supposed to happen at the very end of this function:

```go
	state.ApplyRuntimeStateChanges(runtimeState)
	state.ClearInbox()

	return wf.saveInternalState(ctx, actorID, state)
```

If the inbox isn't getting properly cleared there, then we should do the work to understand why. This loop here is not intended to be used for touching workflow state. Workflow state should only be altered as part of `saveInternalState`.

Reply:
The inbox is getting cleared, however, this line inside ClearInbox() is where the problem resides:
https://github.com/dapr/dapr/blob/c1522bc66ffd7937c2f01f057744d179722a128f/pkg/runtime/wfengine/workflowstate.go#L101

When we create a durable timer, we add it to the inbox: https://github.com/dapr/dapr/blob/c1522bc66ffd7937c2f01f057744d179722a128f/pkg/runtime/wfengine/workflow.go#L312

We then call ClearInbox() later in that runWorkflow() function. This removes the timer from the inbox and increments the s.InboxRemovedCount.

Inside workflowstate.go, we then add state operations: https://github.com/dapr/dapr/blob/c1522bc66ffd7937c2f01f057744d179722a128f/pkg/runtime/wfengine/workflowstate.go#L114

This adds 1 delete operation into the state request since the s.InboxRemovedCount is now 1 and the inbox size is 0: https://github.com/dapr/dapr/blob/c1522bc66ffd7937c2f01f057744d179722a128f/pkg/runtime/wfengine/workflowstate.go#L162

This request is then executed which leads to the 404 since there was never anything actually stored inside the statestore for the timer

Reply:
Wait, does this mean you're adding arbitrary data into the state store so that the delete can succeed? This feels like the wrong direction to me. I understand now why timers are special, and it's because we're adding something to the inbox on line 312 which has no representation in the state store. But if we know that something isn't in the state store, instead of adding it there, we should simply *not try to delete it* - right?

If you agree with the approach, I think we should instead just update `ClearInbox()` to not count `timerFired` events when updating the value of `s.historyRemovedCount`. That way we don't try to delete something that isn't there.
Here's an example (haven't tested it):

```go
func (s *workflowState) ClearInbox() {
	for _, e := range s.Inbox {
		if tf := e.GetTimerFired(); tf != nil {
			// ignore timer events since those aren't saved into the state store
			continue
		}
		s.inboxRemovedCount--
	}
	s.Inbox = nil
	s.inboxAddedCount = 0
}
```